### DISCUSSION TITLE
Remove "realms/{realm}" from endpoints

### DISCUSSION THREAD

**Discussion by y-tabata** [keycloak/keycloak#31505] [Ideas]
### Benefits:
- The URI is simpler
Simply deleting "realms/{realm}" simplifies the endpoint URL and allows us to closely comply with RFC5785, for example. RFC5785 expects "/.well-known/" to be located directly under authority, but Keycloak does not comply with this.
> A well-known URI is a URI [RFC3986] whose path component begins with
> the characters "/.well-known/", and whose scheme is "HTTP", "HTTPS",
> or another scheme that has explicitly been specified to use well-
> known URIs.

- Do not give attackers any hints about how to attack
If "realms/{realm}" is included in the URI, attackers can tell that Keycloak is being used. By removing "realms/{realm}" from the URI, it becomes more difficult for an attacker to easily determine that Keycloak is being used.

### Idea of implementation:
- Set the value set in the frontend URL of each realm as the issuer value of each realm.

|  | "sample" realm's frontend URL | issuer |
| --- | --- | --- |
| **Currently** | https://keycloak.example.com | https://keycloak.example.com/realms/sample |
| **To be** | https://keycloak.example.com | https://keycloak.example.com |

**Comment by ThoreKr** [comment]
How is this supposed to work with multiple realms or how is auto-discovery supposed to work for individual realms then?

**Reply by y-tabata** [reply]
We need to discuss the "how" detail later, we can consider, for example, having mapping information between URLs and realms.

**Reply by darius-m** [reply]
In this case, the "how" is really important, since this is a large change, and it will almost certainly break all (or most) existing OIDC clients and automation scripts (it will likely need to also change the REST API). To me it sounds like an architectural change requires documentation describing the change, how it will be implemented and steps for users to follow to update everything.

**Reply by ThoreKr** [reply]
This. Apart from the point that I think the given current frontend url is incorrect and does not work with multiple realms either.

**Reply by y-tabata** [reply]
As stated in the issue description, I understand that we should be able to discuss how to implement this, under the condition that the URL can be set on a per-realm basis. For existing realms where we want not to remove "realms/{realm}" from the endpoints, we can keep them and there should be minimum impact on existing deployments.
To achieve this, as you say, it may be better to prepare a setting separate from the frontend URL.

**Reply by thomasdarimont** [reply]
I understand the need for this in some scenarios, but I think this should be optional.

However this might actually be a quite involved change because, as mentioned before, a lot of infrastructure, libraries and framework integrations rely on a certain URL structure. However, those integrations are usually adaptable such that users can either override their behaviour or the integration may support the new way over time.

Regarding the URL to realm mapping, there are multiple ways this could be achived. We could introduce a RealmNameResolver SPI which abstracts the way how the realm is resolved from a request. Based on that we could provide users with some options like obtaining the realm name from the request path (as it is now), from a URL Query parameter (non standard for .well-known) or a realm specific subdomain.
For the later we could treat `https://sample.keycloak.com` as `https://keycloak.example.com/realms/sample`.
In cases where users only use one public realm we could allow users to define it as "default" public realm and route requests without a realm reference to the default realm. IMHO think resolving the realm from the request hostname (hostname fragment, or subdomain) is the most straightforward thing to do here combined with a "default public" realm, as users are usually in control of their domains and can control the hostname generation somewhat. However keep in mind that this might introduce an migration problem (existing sessions bound to the old issuer) or you suddenly need to support multiple realms.
Considering the growing support for organisations, the need to have multiple realms might even go away slowly, and the logical separation achived via realms may be achived solely trhough organizations. I think this scenario would also benefit from shorter URLs and a "public default" realm.

Having that said, I see some benefits from supporting a "default realm" with short URLs but I also think this needs some more thoughts regarding the consequences and potential side effects.

**Reply by darius-m** [reply]
@thomasdarimont creating an abstraction makes sense to me, and would likely alleviate the transition issues.

As a side note, there are multiple things to consider if this would be handled using subdomains - Keycloak (or the reverse proxy, depending on the case) would have to be able to handle multiple certificates depending on the subdomain, and that introduces other limitations (e.g., even when using wildcard certificates - which are not things all administrators appreciate, and even some CAs don't -, adding periods in the realm name would break them since `sub.sub.domain.tld` does not match `*.domain.tld`). This likely means that the URL would have to be configurable at realm level.

That being said, I think that this change would be great, especially if a "default" domain would be supported.

**Reply by thomasdarimont** [reply]
Thanks @darius-m yes those exactly the things I was thinking about. Wildcard certificates are one way to tackle the problem with custom subdomains (one level) with the described caveats. Another option its to add multiple SAN (subject alternative names) to the certificate instead of using a wildcard cert, which would allow for "arbitrary" hostnames / custom domains, but again introduces some administrative overhead when a new subdomain needs to be added, but there are at least options. Deriving the realm name from a hostname fragment might also be an option, e.g.: my-realm-auth.keycloak.com, where auth.keycloak.com is the "base domain of Keycloak" and "my-realm-auth.keycloak.com" would need to be an allowed SAN in the certificate.

**Reply by darius-m** [reply]
SANs seem a bit risky, as they they immediately leak the list of domains.
N.B.: Certificates are public for audit purposes regardless, so a determined adversary can still get that information, but it's not as trivial.

**Comment by wadahiro** [comment]
+1 👍 

Based on our experience, we often configure Keycloak with a single realm (excluding the `master` realm). For EIAM, more than 90% of our cases are single realms. In such cases, having the `Issuer Identifier` in the format of `https://example.com/realms/foo` is very strange. The OIDC specification also states at https://openid.net/specs/openid-connect-core-1_0.html#IssuerIdentifier :

> It is RECOMMENDED that only a single Issuer per host be used. However, if a host supports multiple tenants, multiple Issuers for that host may be needed.

This indicates that when using Keycloak as a single tenant, the current `Issuer Identifier` format is not recommended and is undesirable.

Furthermore, even when using multiple realms (e.g., providing customer environments per realm), other IDaaS solutions support custom domain feature for customer tenants. Implementing this in Keycloak with the current implementation results in the following `Issuer Identifier` format, which feels redundant (and doesn't look cool 😢 ):

* `https://foo.example.com/realms/foo`
* `https://bar.example.org/realms/bar`

Therefore, I fully support this proposal.

**Comment by stianst** [comment]
I've wanted something like this for a long time, but never come up with a decent way to implement it.

One challenge is that Keycloak exposes a few resources on the base URL, not just realms:

* /robots.txt - maybe we could remove this, or just always expose it on the base-url
* /realms/{realm}
* /admin - admin can and should be easily configurable to a different hostname, often an internal one
* /resources - realms need this stuff, but could possible be mounted on <realm url>/resources
* /js - we're planning to remove this one

Another is how you would configure the mapping of `URLs` to realms. I've come up with a few different options:

* Allow configuring a URL for a realm (for example https://myrealm.mykeycloak.com) - I'm not too keen on this approach as it would require fetching the mappings from the database, which could be tricky if there are many realms
* Allow configuring virtual URLs of sort for a realm in server-config.
* Make it possible to do this, but only when Keycloak is exposed through a reverse proxy. This might be achievable with simply having a slightly smarter hostname provider 

Another idea was to have the concept of a default realm, so if you don't specify `realms/<realm name>` you'd just get the default realm. 

There's probably more challenges with this as well, but for me it never got past just briefly thinking about it and that it would be nice ;)

**Comment by stianst** [comment]
For this to proceed someone would have to take the initiate on working on a design proposal on how this would all work.

**Comment by slawo** [comment]
I would propose to introduce a realm domain to re-direct realm access from the domain name.
ie
```
master: auth.keycloak.example.com/realm/master/...
master_admin: keycloak.example.com/admin/master/...
realm1: auth.example.com/...
realm1_admin: account.example.com/...
```
It would be possible to use one certificate for the 4 domains, or we could provide an override certificate for realm1

**Comment by ssilvert** [comment]
I will note that there was just a bug fix that used the URL to determine the realm in the case where there is a 404.  It was needed so that the error page text could be localized on a per realm basis.

I consider this to be an edge case and it probably should not be a deciding factor, but wanted to make everyone aware.

**Comment by hentrich-d** [comment]
I would expect a check box right below the frontend url named "relative to realm". Ticking this check box means that the frontend url uniquely reaches to the current realm. Of course, the save button needs a validation that in this case the frontend url is not used for anything else, especially "/realms/master". I would be even fine if the master realm does not allow to tick the box.

I can imagine that the availability of "/resources" or "/robots.txt" needs some extra rounds of thinking.

**Comment by shawkins** [comment]
Marking as closed to move the conversation and refinement to #14623
